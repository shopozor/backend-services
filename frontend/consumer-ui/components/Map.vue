<template>
  <div>
    <client-only>
      <l-map class="mini-map" :zoom="13" :center="position">
        <l-tile-layer url="http://{s}.tile.osm.org/{z}/{x}/{y}.png" />
        <!-- <l-marker :lat-lng="position" :draggable="draggable">
          <l-popup :content="popupContent" />
        </l-marker> -->
      </l-map>
    </client-only>
  </div>
</template>

<script>
export default {
  data: () => ({
    position: [55.607741796855734, 13.018133640289308],
    draggable: true,
    popupContent: 'Sentian HQ'
  })
}
</script>

<style src="leaflet/dist/leaflet.css"></style>
<style>
.mini-map {
  width: 100%;
  height: 100vh !important;
}
</style>
