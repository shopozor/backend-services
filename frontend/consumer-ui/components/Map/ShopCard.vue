<template>
  <div>
    <h1 class="font-extrabold">
      {{ name }}
    </h1>
    {{ description }}
    <!-- TODO: put an image -->
    <!-- TODO: put an address / GPS coordinates -->
  </div>
</template>

<script>
export default {
  props: {
    name: {
      type: String,
      required: true
    },
    description: {
      type: String,
      required: true
    }
    // TODO: id is missing
  }
}
</script>

<!-- style src="leaflet/dist/leaflet.css"></style-->
<style src="~/assets/css/tailwind.css"></style>
