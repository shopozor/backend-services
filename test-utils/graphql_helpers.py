import os.path


def get_query_from_file(graphql_calls_folder, query_name):
    path_to_query = os.path.join(graphql_calls_folder, f'{query_name}.graphql')
    with open(path_to_query, 'r') as myfile:
        return myfile.read().replace('\n', '')
