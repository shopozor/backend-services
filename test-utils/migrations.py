import os
import re
import sh


class HasuraClient:
    def __init__(self, hasura_endpoint):
        self.__cli = sh.Command('hasura')
        self.__endpoint = hasura_endpoint

    def __get_not_present_migrations_count(self, project_folder):
        migrations_folder = os.path.join(project_folder, 'migrations')
        relevant_timestamps = [item.split('_')[0]
                               for item in os.listdir(migrations_folder)]
        statuses = self.__cli('migrate', 'status', '--endpoint', self.__endpoint,
                              '--project', project_folder, '--skip-update-check')
        result = 0
        for timestamp in relevant_timestamps:
            status = re.findall(f'^{timestamp}.*$',
                                str(statuses), re.MULTILINE)[0]
            result += status.count('Not Present')
        return result

    def apply_migrations(self, project_folder):
        result = self.__cli('migrate', 'apply', '--endpoint', self.__endpoint,
                            '--project', project_folder, '--skip-update-check')
        if result.exit_code == 0:
            return self.__get_not_present_migrations_count(project_folder) == 0
        return False

    def __get_number_of_migrations_in_folder(self, migrations_folder):
        return int(len(os.listdir(migrations_folder)) / 2)

    def rollback_migrations(self, project_folder):
        migrations_folder = os.path.join(project_folder, 'migrations')
        nb_migrations = self.__get_number_of_migrations_in_folder(
            migrations_folder)
        result = self.__cli('migrate', 'apply', '--down', nb_migrations, '--endpoint',
                            self.__endpoint, '--project', project_folder, '--skip-update-check')
        if result.exit_code == 0:
            return self.__get_not_present_migrations_count(project_folder) == nb_migrations
        return False
