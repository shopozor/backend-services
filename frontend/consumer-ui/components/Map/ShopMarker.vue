<template>
  <l-marker :lat-lng="position" :icon="icon">
    <l-popup>
      <shop-card :name="shop.name" :description="shop.description" />
    </l-popup>
  </l-marker>
</template>

<script>
import Marker from '~/assets/img/marker.png'
import ValidatedObjectProp from '~/mixins/ValidatedObjectProp'
import ShopCard from '~/components/Map/ShopCard'

export default {
  components: {
    'shop-card': ShopCard
  },
  mixins: [
    ValidatedObjectProp('shop',
      ['latitude', 'longitude', 'name', 'description'])//, 'image']) + address
  ],
  computed: {
    position () {
      return [this.shop.latitude, this.shop.longitude]
    },
    icon () {
      // if we don't disable the following eslint error,
      // then we need to import { L } from 'leaflet'
      // which must not be done on the server-side and is more
      // difficult to setup as just disabling the eslint error
      // eslint-disable-next-line no-undef
      return L.icon({
        iconUrl: Marker
      })
    }
  }
}
</script>
