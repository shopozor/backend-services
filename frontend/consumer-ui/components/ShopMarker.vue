<template>
  <l-marker :lat-lng="position">
    <l-popup>
      <shop-card :name="shop.name" :description="shop.description" />
    </l-popup>
  </l-marker>
</template>

<script>
import ValidatedObjectProp from '../mixins/ValidatedObjectProp'
import ShopCard from './ShopCard'

export default {
  components: {
    'shop-card': ShopCard
  },
  data: () => ({
    popupContent: 'trouduc'
  }),
  computed: {
    position () {
      return [this.shop.latitude, this.shop.longitude]
    }
  },
  mixins: [
    ValidatedObjectProp('shop',
      ['latitude', 'longitude', 'name', 'description'])//, 'image'])
  ]
}
</script>
