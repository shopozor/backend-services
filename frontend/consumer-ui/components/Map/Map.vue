<template>
  <div>
    <client-only>
      <loading :active="!shops" :can-cancel="true" :is-full-page="false" :color="spinnerColor" />
      <l-map class="w-full mini-map" :zoom="zoom" :center="center" :options="options">
        <shop-card v-if="shop" :shop="shop" class="shop-card" />
        <l-tile-layer :url="tilesUrl" />
        <shop-marker v-for="shop in shops" :key="shop.id" :shop="shop" @display-description="onDisplayDescription" />
      </l-map>
    </client-only>
  </div>
</template>

<script>
import ClientOnly from 'vue-client-only'
import Loading from 'vue-loading-overlay'
import ShopCard from '~/components/Map/ShopCard'
import ShopMarker from '~/components/Map/ShopMarker'
import shops from '~graphql/shops'

export default {
  apollo: {
    shops: {
      query: shops
    }
  },
  components: {
    ShopCard,
    ShopMarker,
    ClientOnly,
    Loading
  },
  props: {
    center: {
      type: Array,
      required: true
    },
    zoom: {
      type: Number,
      required: true
    }
  },
  // TODO: test the value of the map's options!
  data: () => ({
    options: {
      gestureHandling: true,
      zoomControl: false
    },
    // TODO: test that shop is initialized with undefined
    shop: undefined,
    spinnerColor: '#e78000ff',
    // cf. https://sosm.ch/projects/tile-service/
    tilesUrl: 'https://tile.osm.ch/osm-swiss-style/{z}/{x}/{y}.png'
  }),
  methods: {
    onDisplayDescription (id) {
      this.shop = this.shops.find(item => item.id === id)
    }
  }
}
</script>

<style src="vue-loading-overlay/dist/vue-loading.css"></style>
<style src="leaflet/dist/leaflet.css"></style>
<style src="leaflet-gesture-handling/dist/leaflet-gesture-handling.css"></style>
<style>
.leaflet-tile-pane {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
}
.shop-card {
  right: 0px;
  top: 0px;
  position: absolute;
  z-index: 999;
}
.mini-map {
  height: 75vh !important;
}
</style>
