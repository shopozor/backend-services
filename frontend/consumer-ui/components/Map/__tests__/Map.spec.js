import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import { LMap, LMarker, LTileLayer } from 'vue2-leaflet'
// eslint-disable-next-line no-unused-vars
import { GestureHandling } from 'leaflet-gesture-handling'
import Map from '../Map'
import ShopMarker from '../ShopMarker'
import ShopsData from '~fixtures/Consumer/Shops'

function getLocalVue () {
  const localVue = createLocalVue()
  localVue.component('l-map', LMap)
  localVue.component('l-tile-layer', LTileLayer)
  localVue.component('l-marker', LMarker)
  return localVue
}

describe('Map', () => {
  const center = [46.718852, 7.097669]
  const zoom = 11
  const localVue = getLocalVue()

  it('indicates loading shops', async () => {
    // Given I have a loading map
    const wrapper = mount(Map, {
      localVue,
      propsData: {
        center,
        zoom
      },
      mocks: {
        $apollo: {
          queries: {
            shops: {
              loading: true
            }
          }
        }
      }
    })

    // Then it is overlayed by a loading spinner or something equivalent
    await localVue.nextTick()
    const overlayElement = wrapper.find('.vld-overlay')
    expect(overlayElement.isVisible()).toBeTruthy()
  })

  it('is initialized with gesture handling', () => {
    // Given I have a loaded map
    const wrapper = mount(Map, {
      localVue,
      mocks: {
        $apollo: {
          queries: {
            shops: {
              loading: false
            }
          }
        }
      },
      propsData: {
        center,
        zoom
      }
    })

    // Then the gesture handling is enabled
    const leafletMap = wrapper.find('.vue2leaflet-map')
    const attributes = leafletMap.attributes()
    const expectedGestureHandlingAttributes = ['data-gesture-handling-touch-content', 'data-gesture-handling-scroll-content']
    const hasGestureHandling = expectedGestureHandlingAttributes.some(r => Object.keys(attributes).includes(r))
    expect(hasGestureHandling).toBeTruthy()
  })

  it('does not show zoom control', () => {
    const wrapper = mount(Map, {
      localVue,
      propsData: {
        center,
        zoom
      },
      mocks: {
        $apollo: {
          queries: {
            shops: {
              loading: false
            }
          }
        }
      }
    })
    const zoomControl = wrapper.find('.leaflet-control-zoom')
    expect(zoomControl.exists()).toBeFalsy()
  })

  it('is initialized with no shop description popup', () => {
    const wrapper = shallowMount(Map, {
      localVue,
      propsData: {
        center,
        zoom
      },
      mocks: {
        $apollo: {
          queries: {
            shops: {
              loading: false
            }
          }
        }
      }
    })
    expect(wrapper.vm.shop).toBeUndefined()
  })

  it('displays the selected shop description', () => {
    // Given I have shops data and am not loading server data
    const wrapper = mount(Map, {
      localVue,
      propsData: {
        center,
        zoom
      },
      data: () => ({
        shops: ShopsData.data.shops
      }),
      mocks: {
        $apollo: {
          queries: {
            shops: {
              loading: false
            }
          }
        }
      }
    })
    expect(wrapper.vm.shop).toBeUndefined()

    // When a child marker emits the display-description event
    const expectedShop = ShopsData.data.shops[0]
    const marker = wrapper.find(ShopMarker)
    marker.vm.$emit('display-description', expectedShop.id)

    // Then a shop is selected to display its description
    expect(wrapper.vm.shop).toBe(expectedShop)
  })

  it('clears shop description popup upon clicking the map', () => {
    // Given I have a shop selected for description
    const wrapper = mount(Map, {
      localVue,
      propsData: {
        center,
        zoom
      },
      data: () => ({
        shops: ShopsData.data.shops,
        shop: ShopsData.data.shops[0]
      }),
      mocks: {
        $apollo: {
          queries: {
            shops: {
              loading: false
            }
          }
        }
      }
    })
    expect(wrapper.vm.shop).toBeDefined()

    // When I click the map
    const map = wrapper.find(LMap)
    map.trigger('click')

    // Then I have no shop selected anymore
    expect(wrapper.vm.shop).toBeUndefined()
  })
})
