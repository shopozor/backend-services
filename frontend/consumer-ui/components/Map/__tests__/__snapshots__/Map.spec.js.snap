// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders spinner when loading shops 1`] = `
<div>
  <div
    aria-busy="true"
    aria-label="Loading"
    class="vld-overlay is-active z-999"
    name="fade"
    tabindex="0"
  >
    <div
      class="vld-background"
    />
    <div
      class="vld-icon"
    >
      <svg
        height="64"
        stroke="#e78000ff"
        viewBox="0 0 38 38"
        width="64"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <g
            stroke-width="2"
            transform="translate(1 1)"
          >
            <circle
              cx="18"
              cy="18"
              r="18"
              stroke-opacity=".25"
            />
            <path
              d="M36 18c0-9.94-8.06-18-18-18"
            >
              <animatetransform
                attributeName="transform"
                dur="0.8s"
                from="0 18 18"
                repeatCount="indefinite"
                to="360 18 18"
                type="rotate"
              />
            </path>
          </g>
        </g>
      </svg>
    </div>
  </div>
   
  <div
    class="vue2leaflet-map w-full fill-partial-height leaflet-container"
    data-gesture-handling-scroll-content="CTRL + scroll pour zoomer"
    data-gesture-handling-touch-content="Utiliser 2 doigts pour bouger la carte"
    style="position: relative;"
    tabindex="0"
  >
    <!---->
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://tile.osm.ch/switzerland/11/1064/722.png"
              style="width: 256px; height: 256px; left: -97px; top: -173px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      />
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
        </div>
      </div>
    </div>
     
    <div />
     
  </div>
</div>
`;
